### PR TITLE
SLCORE-763 Delete bindings of connection deleted

### PR DESCRIPTION
**Pull request by vojtech-suchy-sonarsource**
# For SonarSourcers:

- [ ] Prefix the commit message with the ticket number, i.e. `SLCORE-XXXX` (or use `NO-JIRA` for a trivial change not tracked by a ticket)
- [ ] When changing an API:
    - [ ] Explain in the JavaDoc the purpose of the new API
    - [ ] Document the change in [API_CHANGES.md](https://github.com/SonarSource/sonarlint-core/blob/master/API_CHANGES.md)
    - [ ] If the change breaks the current API, explicitly communicate those to the impacted consumers prior to merging (eg. IDE squad)
- [ ] Make sure the tests adhere to the convention:
    - [ ] All test method names should use `snake_case`, for example: `test_validate_input`.
- [ ] Make sure checks are green: build passes, Quality Gate is green

### CODE REVIEW

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/BindingConfiguration.java [hunk 1]
```diff
@@ -37,6 +37,10 @@ public BindingConfiguration(@Nullable String connectionId, @Nullable String sona
     this.bindingSuggestionDisabled = bindingSuggestionDisabled;
   }
 
+  public static BindingConfiguration noBinding() {
+    return new BindingConfiguration(null, null, false);
+  }
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/BindingConfiguration.java:42]: :+1:

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/ConfigurationService.java [hunk 2]
```diff
@@ -100,6 +102,13 @@ public void didUpdateBinding(String configScopeId, BindingConfigurationDto updat
     }
   }
 
+  @EventListener
+  public void connectionRemoved(ConnectionConfigurationRemovedEvent event) {
+    var bindingConfigurationByConfigScope = repository.removeBindingForConnection(event.getRemovedConnectionId());
+    bindingConfigurationByConfigScope.forEach((configScope, bindingConfiguration) ->
+      applicationEventPublisher.publishEvent(new BindingConfigChangedEvent(configScope, bindingConfiguration, BindingConfiguration.noBinding())));
+  }
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/ConfigurationService.java:110]: When we automatically unbind the scope, we should still keep the value that is set in `BindingConfiguration.bindingSuggestionDisabled`. This is a user setting, and we should keep it as it is even if we force the unbind. This flag should maybe be outside of BindingConfiguration, we have [SLCORE-752](https://sonarsource.atlassian.net/browse/SLCORE-752) for that

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java [hunk 3]
```diff
@@ -62,6 +63,19 @@ public ConfigurationScopeWithBinding remove(String idToRemove) {
     return removedScope == null ? null : new ConfigurationScopeWithBinding(removedScope, removeBindingConfiguration);
   }
 
+  public Map<String, BindingConfiguration> removeBindingForConnection(String connectionId) {
+    var removedBindingByConfigScope = new HashMap<String, BindingConfiguration>();
+    var configScopesToUnbind =
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java:68]: Small suggestion: `configScopesToUnbind` -> `configScopeIdsToUnbind`

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java [hunk 4]
```diff
@@ -62,6 +63,19 @@ public ConfigurationScopeWithBinding remove(String idToRemove) {
...
+      bindingPerConfigScopeId.entrySet().stream().filter(e -> connectionId.equals(e.getValue().getConnectionId())).map(Map.Entry::getKey).collect(toSet());
+    configScopesToUnbind.forEach(configScope -> {
```
Comment (damien-urruty-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java:70]: Similar: `configScope` -> `configScopeId`

**Review by damien-urruty-sonarsource — APPROVED** [review]
Minor comments, LGTM!

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java (hunk 4 above)
Comment (vojtech-suchy-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java]: Addressed in [d9d5ad2](https://github.com/SonarSource/sonarlint-core/pull/982/commits/d9d5ad261521e1be145998b0fa925f68b0981f2a)

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java (hunk 3 above)
Comment (vojtech-suchy-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/repository/config/ConfigurationRepository.java]: Addressed in [d9d5ad2](https://github.com/SonarSource/sonarlint-core/pull/982/commits/d9d5ad261521e1be145998b0fa925f68b0981f2a)

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/ConfigurationService.java (hunk 2 above)
Comment (vojtech-suchy-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/ConfigurationService.java]: Addressed in [d9d5ad2](https://github.com/SonarSource/sonarlint-core/pull/982/commits/d9d5ad261521e1be145998b0fa925f68b0981f2a)

**Pull request merged**